multinode/adapter.go

```diff
@@ -0,0 +1,302 @@
+package multinode
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+)
+
+type AdapterConfig interface {
+	NewHeadsPollInterval() time.Duration
+	FinalizedBlockPollInterval() time.Duration
+}
+
+// Adapter is used to integrate multinode into chain-specific clients
+type Adapter[RPC any, HEAD Head] struct {
+	cfg         AdapterConfig
+	log         logger.Logger
+	rpc         *RPC
+	ctxTimeout  time.Duration
+	stateMu     sync.RWMutex // protects state* fields
+	subsSliceMu sync.RWMutex
+	subs        map[Subscription]struct{}
+
+	latestBlock          func(ctx context.Context, rpc *RPC) (HEAD, error)
+	latestFinalizedBlock func(ctx context.Context, rpc *RPC) (HEAD, error)
+
+	// chStopInFlight can be closed to immediately cancel all in-flight requests on
+	// this RpcMultiNodeAdapter. Closing and replacing should be serialized through
+	// stateMu since it can happen on state transitions as well as RpcMultiNodeAdapter Close.
+	chStopInFlight chan struct{}
```

Comment:
Should we rename this to a lifeCycleCh to signal that it's closed once RPC is declared unhealthy?